// @sentry/server-utils/build/esm/orchestrion/index.js
import { amqplibChannelIntegration } from '../integrations/tracing-channel/amqplib.js';
import { anthropicChannelIntegration } from '../integrations/tracing-channel/anthropic.js';
import { googleGenAIChannelIntegration } from '../integrations/tracing-channel/google-genai.js';
import { graphqlDiagnosticsChannelIntegration } from '../integrations/tracing-channel/graphql/index.js';
export { graphqlChannelIntegration } from '../integrations/tracing-channel/graphql/index.js';
import { hapiChannelIntegration } from '../integrations/tracing-channel/hapi.js';
export { ioredisChannelIntegration } from '../integrations/tracing-channel/ioredis.js';
import { lruMemoizerChannelIntegration } from '../integrations/tracing-channel/lru-memoizer.js';
import { mysqlChannelIntegration } from '../integrations/tracing-channel/mysql.js';
import { openaiChannelIntegration } from '../integrations/tracing-channel/openai.js';
import { postgresChannelIntegration } from '../integrations/tracing-channel/postgres.js';
import { postgresJsChannelIntegration } from '../integrations/tracing-channel/postgres-js.js';
import { vercelAiChannelIntegration } from '../integrations/tracing-channel/vercel-ai.js';
import { expressChannelIntegration } from '../integrations/tracing-channel/express/index.js';
export { detectOrchestrionSetup, isOrchestrionInjected } from './detect.js';
export { redisChannelIntegration } from '../integrations/tracing-channel/redis.js';

const channelIntegrations = {
  postgresIntegration: postgresChannelIntegration,
  postgresJsIntegration: postgresJsChannelIntegration,
  mysqlIntegration: mysqlChannelIntegration,
  lruMemoizerIntegration: lruMemoizerChannelIntegration,
  openaiIntegration: openaiChannelIntegration,
  anthropicIntegration: anthropicChannelIntegration,
  googleGenAIIntegration: googleGenAIChannelIntegration,
  vercelAiIntegration: vercelAiChannelIntegration,
  amqplibIntegration: amqplibChannelIntegration,
  hapiIntegration: hapiChannelIntegration,
  expressIntegration: expressChannelIntegration,
  graphqlIntegration: graphqlDiagnosticsChannelIntegration
};

export { amqplibChannelIntegration, anthropicChannelIntegration, channelIntegrations, expressChannelIntegration, googleGenAIChannelIntegration, hapiChannelIntegration, lruMemoizerChannelIntegration, mysqlChannelIntegration, openaiChannelIntegration, postgresChannelIntegration, postgresJsChannelIntegration, vercelAiChannelIntegration };
